### PR TITLE
Ngtsc/ttc/strict

### PR DESCRIPTION
**Pull request by alxhub**
Takeover of https://github.com/angular/angular/pull/33273 for faster round trips.

### CODE REVIEW

**Comment by alxhub** [comment]
[Presubmit](https://test.corp.google.com/ui#id=OCL:276489192:BASE:276489222:1571929591524:c92d3e19)

File: packages/compiler-cli/src/ngtsc/typecheck/src/api.ts [hunk 1]
```diff
@@ -133,14 +148,24 @@ export interface TypeCheckingConfig {
    */
   checkTypeOfDomEvents: boolean;
 
+  /**
+   * Whether to infer the type of local references to DOM elements.
+   *
+   * If this is `true`, the type of a `#ref` variable on a DOM in the template will be determined by
```
Comment (kara) [inline packages/compiler-cli/src/ngtsc/typecheck/src/api.ts]:
```suggestion
   * If this is `true`, the type of a `#ref` variable on a DOM node in the template will be determined by
```

File: packages/compiler-cli/src/transformers/api.ts [hunk 2]
```diff
@@ -108,10 +108,111 @@ export interface CompilerOptions extends ts.CompilerOptions {
...
+   * Whether to use strict null types for input bindings for directives.
+   *
+   * If this is `true`, applications that are compiled with TypeScript's `strictNullChecks` enabled
+   * will produce type errors for bindings which can evaluate to `undefined` or `null` where the
+   * inputs's type does not include `undefined` or `null` in its type. If set to `false`, all
+   * binding expressions are wrapped in a non-null assertion operator to effectively disable strict
+   * null checks.
+   *
+   * Defaults to `false`, even if "fullTemplateTypeCheck" is set. Note that if `strictInputTypes` is
+   * not set, or set to `false`, this flag has no effect.
```
Comment (kara) [inline packages/compiler-cli/src/transformers/api.ts]: All these comments are great 👍

File: packages/compiler-cli/src/transformers/api.ts [hunk 3]
```diff
@@ -108,10 +108,111 @@ export interface CompilerOptions extends ts.CompilerOptions {
...
+   * expression `a != null ? a.b : a`.
+   *
+   * Defaults to `false`, even if "fullTemplateTypeCheck" is set.
+   */
+  strictSafeNavigationTypes?: boolean;
+
+  /**
+   * Whether to infer the type of local references.
+   *
+   * If this is `true`, the type of a `#ref` variable on a DOM in the template will be determined by
```
Comment (kara) [inline packages/compiler-cli/src/transformers/api.ts]:
```suggestion
   * If this is `true`, the type of a `#ref` variable on a DOM node in the template will be determined by
```

File: packages/compiler-cli/src/transformers/api.ts [hunk 4]
```diff
@@ -108,10 +108,111 @@ export interface CompilerOptions extends ts.CompilerOptions {
...
+   * the type of `document.createElement` for the given DOM node. If set to `false`, the type of
+   * `ref` for DOM nodes will be `any`.
+   *
+   * Defaults to `false`, even if "fullTemplateTypeCheck" is set.
+   */
+  strictDomLocalRefTypes?: boolean;
+
+  /**
+   * Whether to infer the type of the `$event` variable in event bindings for directive outputs.
```
Comment (kara) [inline packages/compiler-cli/src/transformers/api.ts]: Should we note here that it encompasses animation outputs too?

**Review by kara — APPROVED** [review]
LGTM, some minor nits in comments

File: packages/compiler-cli/src/transformers/api.ts (hunk 4 above)
Comment (alxhub) [reply packages/compiler-cli/src/transformers/api.ts]: Done.

**Pull request closed without merging**